examples/common/imgui_ui/ui.cpp

```diff
@@ -0,0 +1,198 @@
+
+    // Setup Platform/Renderer backends
+    ImGui_ImplSDL2_InitForOpenGL(*window, *gl_context);
+    ImGui_ImplOpenGL3_Init();
+}
+
+void UiShutdown(SDL_GLContext * gl_context, SDL_Window ** window)
+{
+    ImGui_ImplOpenGL3_Shutdown();
+    ImGui_ImplSDL2_Shutdown();
+    ImGui::DestroyContext();
+
+    SDL_GL_DeleteContext(*gl_context);
+    SDL_DestroyWindow(*window);
+    SDL_Quit();
+}
+
+void EventLoop(ImguiUi * ui)
+{
+    gUiRunning = true;
+    SDL_GLContext gl_context;
+    SDL_Window * window = nullptr;
+
+    UiInit(&gl_context, &window);
+
+    ImGuiIO & io = ImGui::GetIO();
+
+    while (gUiRunning.load())
+    {
+        SDL_Event event;
+        while (SDL_PollEvent(&event))
+        {
+            ImGui_ImplSDL2_ProcessEvent(&event);
+            if ((event.type == SDL_QUIT) ||
+                (event.type == SDL_WINDOWEVENT && event.window.event == SDL_WINDOWEVENT_CLOSE &&
+                 event.window.windowID == SDL_GetWindowID(window)))
+            {
+                gUiRunning = false;
+            }
+        }
+
+        ImGui_ImplOpenGL3_NewFrame();
+        ImGui_ImplSDL2_NewFrame();
+        ImGui::NewFrame();
+
+        ui->UpdateState();
+        ui->Render();
+
+        // rendering
+        ImGui::Render();
+        glViewport(0, 0, (int) io.DisplaySize.x, (int) io.DisplaySize.y);
+        glClearColor(0, 0, 0, 0);
+        glClear(GL_COLOR_BUFFER_BIT);
+        ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
+        SDL_GL_SwapWindow(window);
+    }
+
+    UiShutdown(&gl_context, &window);
+
+    ChipLogProgress(AppServer, "UI thread Stopped...");
+}
+
+} // namespace
+
+void ImguiUi::RunMainLoop()
+{
+    // Guaranteed to be on the main task (no chip event loop started yet)
+    ChipLoopLoadInitialState();
+
+    // Platform event loop will be on a separate thread,
+    // while the event UI loop will be on the main thread.
+    chip::DeviceLayer::PlatformMgr().StartEventLoopTask();
+
+    // StopEventLoop will stop the loop below. It is called
+    // from within SignalSafeStopMainLoop below and
+    // UI knows how to stop itself if windows are closed.
+    EventLoop(this);
+
+    // Stop the chip main loop as well. This is expected to
+    // wait for the task to finish.
+    chip::DeviceLayer::PlatformMgr().StopEventLoopTask();
+}
+
+void ImguiUi::SignalSafeStopMainLoop()
+{
+    gUiRunning = false;
+}
+
+void ImguiUi::ChipLoopStateUpdate()
+{
+    assertChipStackLockedByCurrentThread();
+    for (auto it = mWindows.begin(); it != mWindows.end(); it++)
+    {
+        (*it)->UpdateState();
+    }
+}
+
+void ImguiUi::ChipLoopLoadInitialState()
+{
+    for (auto it = mWindows.begin(); it != mWindows.end(); it++)
+    {
+        (*it)->LoadInitialState();
+    }
+}
+
+void ImguiUi::Render()
+{
+    for (auto it = mWindows.begin(); it != mWindows.end(); it++)
+    {
+        (*it)->Render();
+    }
+}
+
+void ImguiUi::UpdateState()
+{
+    chip::DeviceLayer::StackLock lock;
```

Comment:
This is not going to work on Darwin, fwiw; not sure whether it's meant to.

Reply:
I thought this was the clearest way instead of the `ScheduleWork` I had before.

However if this is not working on darwin I can revert to schedule and semaphore. 
Any reason it does not work? I thought we had a `stackLock` that is an alternative to running on the same thread.

Should we make compilation fail if this is used in darwin?

Reply:
Updated to ScheduleWork and semaphores. Longer code :(